### PR TITLE
Updates unit tests; `ValidatorsMethod`

### LINKED ISSUE #8115: Unit tests review for validators module (LIP 0044)
The tests in [this folder](https://github.com/LiskHQ/lisk-sdk/tree/development/framework/test/unit/modules/validators) were reviewed.

# Endpoints

## validateBLSKey

- for the test for [existing BLS key](https://github.com/LiskHQ/lisk-sdk/blob/5e6afb58f4655ac19ca832a2c203bd5418679540/framework/test/unit/modules/validators/endpoint.spec.ts#L44), a valid proof of possession should be used for the given BLS key. Otherwise, the function may return `false` due to invalid proof of possession and not because the key exists already

# Methods

## registerValidatorKeys

- For some tests for which the function is expected to fail, it is tested that the substores does not contain the corresponding entries afterwards. For example [here](https://github.com/LiskHQ/lisk-sdk/blob/5e6afb58f4655ac19ca832a2c203bd5418679540/framework/test/unit/modules/validators/method.spec.ts#L239-L240). For some other tests, (for example [here](https://github.com/LiskHQ/lisk-sdk/blob/5e6afb58f4655ac19ca832a2c203bd5418679540/framework/test/unit/modules/validators/method.spec.ts#L147)) it is not done. It would be good to do it for every test where it is expected to fail.

## setValidatorBLSKey

- [This test](https://github.com/LiskHQ/lisk-sdk/blob/5e6afb58f4655ac19ca832a2c203bd5418679540/framework/test/unit/modules/validators/method.spec.ts#L378) should be improved by using a different BLS key in the [validator account](https://github.com/LiskHQ/lisk-sdk/blob/5e6afb58f4655ac19ca832a2c203bd5418679540/framework/test/unit/modules/validators/method.spec.ts#L386). Otherwise, it is not clear if the test should fail because the BLS key exists in the corresponding validator account in the validators substore or in the registered bls keys substore.
- We should have a test where a validator account with a valid BLS key is already stored in the validators substore, and `setValidatorBLSKey` is called for this account. In contrast to an old version of the LIP, this should NOT fail.
- For tests where the BLS keys have [incorrect length](https://github.com/LiskHQ/lisk-sdk/blob/5e6afb58f4655ac19ca832a2c203bd5418679540/framework/test/unit/modules/validators/method.spec.ts#L463-L485), it would be good to set up a validator account with the given address in the validators substore. Otherwise, it may just fail because the account does not exist.

## setValidatorGeneratorKey

- It would be good to have a test like [this one](https://github.com/LiskHQ/lisk-sdk/blob/5e6afb58f4655ac19ca832a2c203bd5418679540/framework/test/unit/modules/validators/method.spec.ts#L493), but where the new generator key is different from the existing one. And then it should be tested that the generator key in the substore was indeed updated. The existing test, where the new and the old generator key are equal, should be kept though.
- For the [tests where the generator key has an incorrect length](https://github.com/LiskHQ/lisk-sdk/blob/5e6afb58f4655ac19ca832a2c203bd5418679540/framework/test/unit/modules/validators/method.spec.ts#L564-L584), it would be good to set up the validator account with corresponding address before. Otherwise, it may just fail because the account does not exist.

## getGeneratorsBetweenTimestamps

- For [this test](https://github.com/LiskHQ/lisk-sdk/blob/5e6afb58f4655ac19ca832a2c203bd5418679540/framework/test/unit/modules/validators/method.spec.ts#L634):
  - It must also work for even stricter inputs, e.g. lowering the [endTimestamp](https://github.com/LiskHQ/lisk-sdk/blob/5e6afb58f4655ac19ca832a2c203bd5418679540/framework/test/unit/modules/validators/method.spec.ts#L641) to `timePerRound + 2 * blockTime`.
  - The desciption "if ... difference between input timestamps is greater than one round" is not correct. It must be "if difference between timestamps is larger than or equal to the length of one round plus two block slots."
- For [this test](https://github.com/LiskHQ/lisk-sdk/blob/5e6afb58f4655ac19ca832a2c203bd5418679540/framework/test/unit/modules/validators/method.spec.ts#L653):
  - It must also work for even stricter inputs, e.g. lowering the [endTimestamp](https://github.com/LiskHQ/lisk-sdk/blob/5e6afb58f4655ac19ca832a2c203bd5418679540/framework/test/unit/modules/validators/method.spec.ts#L660) to `timePerRound * 2 + 2 * blockTime`.
  - The desciption "if ... difference between input timestamps is greater than 2 rounds" is not correct. It must be "if difference between timestamps is larger than or equal to the length of two rounds plus two block slots."
- The description of [this test](https://github.com/LiskHQ/lisk-sdk/blob/5e6afb58f4655ac19ca832a2c203bd5418679540/framework/test/unit/modules/validators/method.spec.ts#L723) is wrong because the timestamps are equal but not `startSlotNumber` and `endSlotNumber`
- it would be nice to have an additional test that checks which subset of validators is counted and which one is not. For example:
  - `startTimestamp = 99 * blockTime`
  - `endTimestamp = 103 * blockTime`
  - expected result:  the result contains the 3 entries `generatorList[99]`, `generatorList[100]` and `generatorList[0]` with count 1 for each entry
  
## getValidatorKeys

- Maybe not relevant for the tests, but the generator key should have 32 bytes and the BLS key 48 bytes, and not the [other way around](https://github.com/LiskHQ/lisk-sdk/blob/5e6afb58f4655ac19ca832a2c203bd5418679540/framework/test/unit/modules/validators/method.spec.ts#L737-L738). 
  
## registerValidatorWithoutBLSKey

- in [this test](https://github.com/LiskHQ/lisk-sdk/blob/5e6afb58f4655ac19ca832a2c203bd5418679540/framework/test/unit/modules/validators/method.spec.ts#L802), it should also be checked that `setValidatorAccount.blsKey` equals `INVALID_BLS_KEY`

## setValidatorParams

There are no tests yet for this function. It should at least be tested that
-  the _validators params substore_ is updated correctly, and that
- the params are forwarded to the consensus domain

### PR DESCRIPTION
**Pull request by has5aan**
### What was the problem?

This PR resolves issue #8115 & issue #7267  Some points in the issue, potentially requiring no change are responded to in a comment on the issue.

### How was it solved?

Adds and updates tests for `ValidatorsMethod`.

### How was it tested?
All tests are passing.

### CODE REVIEW

File: framework/test/unit/modules/validators/method.spec.ts [hunk 1]
```diff
@@ -632,14 +673,14 @@ describe('ValidatorsModuleMethod', () => {
 			).resolves.toBeObject();
 		});
 
-		it('should be able to return generators with at least one generator assigned more than one slot if input timestamps are valid and difference between input timestamps is greater than one round', async () => {
+		it('should be able to return generators with at least one generator assigned more than one slot if input timestamps are valid and difference between input timestamps is greater than one round plus two blocks', async () => {
```
Comment (AndreasKendziorra) [inline framework/test/unit/modules/validators/method.spec.ts:676]:
```suggestion
		it('should be able to return generators with at least one generator assigned more than one slot if input timestamps are valid and difference between input timestamps is greater than or equal to one round plus two blocks', async () => {
```

File: framework/test/unit/modules/validators/method.spec.ts [hunk 2]
```diff
@@ -721,11 +762,37 @@ describe('ValidatorsModuleMethod', () => {
 			).resolves.toEqual({});
 		});
 
-		it('should return empty result when startSlotNumber equals endSlotNumber but in the same block slot', async () => {
+		it('should return empty result when startTimestamp equals endTimestamp but in the same block slot', async () => {
```
Comment (AndreasKendziorra) [inline framework/test/unit/modules/validators/method.spec.ts:765]:
```suggestion
		it('should return empty result when startTimestamp equals endTimestamp', async () => {
```

**Review by AndreasKendziorra — CHANGES_REQUESTED** [review]
I left an additional comment in issue #8115

File: framework/test/unit/modules/validators/method.spec.ts [hunk 3]
```diff
@@ -351,6 +366,31 @@ describe('ValidatorsModuleMethod', () => {
 			);
 		});
 
+		it('should not be able to correctly set bls key for validator if address exists with valid blsKey, key is not registered and proof of possession is valid', async () => {
```
[the remark this answers is not in the document]
Comment (has5aan) [reply framework/test/unit/modules/validators/method.spec.ts:369]: Should the entry in BLSKeyStore against the existing valid blsKey of the validator be removed?

[the remark this answers is not in the document]
Comment (AndreasKendziorra) [reply framework/test/unit/modules/validators/method.spec.ts:369]: No, the existing blsKey should not be removed from BLSKeyStore.

[the remark this answers is not in the document]
Comment (has5aan) [reply framework/test/unit/modules/validators/method.spec.ts]: Updates [here](https://github.com/LiskHQ/lisk-sdk/pull/9126/commits/a8df3f192102af5a713882eeb15b7fbc3e79a78b).

**Review by AndreasKendziorra — APPROVED** [review]

File: framework/src/modules/validators/method.ts [hunk 4]
```diff
@@ -208,9 +208,6 @@ export class ValidatorsMethod extends BaseMethod {
 		}
 
 		const validatorAccount = await validatorsSubStore.get(methodContext, validatorAddress);
-		if (!validatorAccount.blsKey.equals(INVALID_BLS_KEY)) {
-			return false;
-		}
```
Comment (shuse2) [inline framework/src/modules/validators/method.ts:213]: why is this removed?

Comment (AndreasKendziorra) [reply framework/src/modules/validators/method.ts:213]: At some point, the requirement of disallowing updates of the BLS key was dropped.

Comment (AndreasKendziorra) [reply framework/src/modules/validators/method.ts:213]: hm. Then it is not aligned with the specs in the LIP, but on the other hand it is not critical, because we also provide this [warning](https://github.com/LiskHQ/lips/blob/8c374e7649c48bcc8bf9015ca2ebc8272b72998d/proposals/lip-0044.md?plain=1#L426):

> Note: For blockchains build with the SDK v6, this function should be not be called except from the [Legacy module](https://github.com/LiskHQ/lips/blob/main/proposals/lip-0050.md#register-keys) used on the Lisk Mainchain. The reason is that changing BLS keys is not supported by the (interoperability) protocol.
